imagetest/testworkflow.go

```diff
@@ -0,0 +1,245 @@
+func PrintTests(ctx context.Context, testWorkflows []*TestWorkflow, project, zone string) {
+	finalizeWorkflows(testWorkflows, zone, project)
+	for _, test := range testWorkflows {
+		if test.wf == nil {
+			continue
+		}
+		test.wf.Print(ctx)
+	}
+}
+
+// ValidateTests validates all test workflows.
+func ValidateTests(ctx context.Context, testWorkflows []*TestWorkflow, project, zone string) error {
+	finalizeWorkflows(testWorkflows, zone, project)
+	for _, test := range testWorkflows {
+		if test.wf == nil {
+			continue
+		}
+		if err := test.wf.Validate(ctx); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// RunTests runs all test workflows.
+func RunTests(ctx context.Context, testWorkflows []*TestWorkflow, outPath, project, zone string, parallelCount int) {
+	var err error
+	client, err = storage.NewClient(ctx)
+	if err != nil {
+		log.Fatalf("Failed to set up storage client: %v", err)
+	}
+	finalizeWorkflows(testWorkflows, zone, project)
+
+	testResults := make(chan testResult, len(testWorkflows))
+	testchan := make(chan *TestWorkflow, len(testWorkflows))
+
+	var wg sync.WaitGroup
+	for i := 0; i < parallelCount; i++ {
+		wg.Add(1)
+		go func(id int) {
+			defer wg.Done()
+			for test := range testchan {
+				testResults <- runTestWorkflow(ctx, test)
+			}
+		}(i)
+	}
+	for _, ts := range testWorkflows {
+		testchan <- ts
+	}
+	close(testchan)
+	wg.Wait()
+
+	var suites testSuites
+	for i := 0; i < len(testWorkflows); i++ {
+		suites.TestSuite = append(suites.TestSuite, parseResult(<-testResults))
+	}
+
+	bytes, err := xml.MarshalIndent(suites, "", "\t")
+	if err != nil {
+		fmt.Printf("failed to marshall result: %v\n", err)
+		return
+	}
+	outFile, err := os.Create(outPath)
+	if err != nil {
+		fmt.Printf("failed to create output file: %v\n", err)
+		return
+	}
+	defer outFile.Close()
+
+	outFile.Write(bytes)
+	outFile.Write([]byte{'\n'})
+
+	return
+}
+
+func runTestWorkflow(ctx context.Context, test *TestWorkflow) testResult {
+	var res testResult
+	res.testWorkflow = test
+	if test.skipped {
+		res.skipped = true
+		res.err = fmt.Errorf("Test suite was skipped")
+		return res
+	}
+	fmt.Printf("runTestWorkflow: running %s on %s (ID %s)\n", test.Name, test.ShortImage, test.wf.ID())
+	if err := test.wf.Run(ctx); err != nil {
+		res.err = err
+		return res
+	}
+	results, err := getTestResults(ctx, test)
+	if err != nil {
+		res.err = err
+		return res
+	}
+	res.results = results
+	res.workflowSuccess = true
+
+	return res
+}
+
+// gets result struct and converts to a jUnit TestSuite
+func parseResult(res testResult) *testSuite {
+	ret := &testSuite{}
+
+	switch {
+	case res.skipped:
+		ret.Tests = 1
+		ret.Skipped = 1
+	case res.workflowSuccess:
+		// Workflow completed without error. Only in this case do we try to parse the result.
+		ret = convertToTestSuite(res.results)
+	default:
+		ret.Tests = 1
+		ret.Errors = 1
+		if res.err != nil {
+			ret.SystemErr = res.err.Error()
+		} else {
+			ret.SystemErr = "Unknown status"
```

Comment:
Can this happen? Looking at `runTestWorkflow` it seems either `err` is being set or `workflowSuccess` is True